### PR TITLE
Exclude empty item from completion list

### PR DESCRIPTION
**Pull request by genlu**
Fix https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1152109

In this case, SemanticModel would return a local symbol named `` of type `int`, we want to exclude that from completion list.
```cs
static void Main()
{
    int$$
}
```

@jasonmalinowski not sure what's the reason for this compiler behavior, should the fix be in compiler instead?

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 1]
```diff
@@ -10733,5 +10733,23 @@ static void M()
...
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(1152109, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1152109")]
+        public async Task NoItemWithEmptyDisplayName()
+        {
+            var markup = @"
+class C
+{
+    static void M()
+    {
+        int$$
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:10746]: Shouldn't a space exist between `int` and `$$`?

```suggestion
        int $$
```

**Comment by Youssef1313** [comment]
> not sure what's the reason for this compiler behavior, should the fix be in compiler instead?

I think the semantic model shouldn't return an empty-named symbol. I think the proper fix might be in the compiler-side, not the IDE-side.

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs [hunk 2]
```diff
@@ -134,6 +134,7 @@ private static bool IsTypeImplicitlyConvertible(Compilation compilation, ITypeSy
         {
             var symbolGroups = from symbol in symbols
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:135]: you might want to ensure that hte code that produces these symbols does a .CanBeReferencedByName on them.

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs [hunk 3]
```diff
@@ -134,6 +134,7 @@ private static bool IsTypeImplicitlyConvertible(Compilation compilation, ITypeSy
...
                                let texts = GetDisplayAndSuffixAndInsertionText(symbol, contextLookup(symbol))
+                               where !string.IsNullOrWhiteSpace(texts.displayText)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:137]: see .CanBeReferencedByName instead of this.

**Comment by CyrusNajmabadi** [comment]
> I think the semantic model shouldn't return an empty-named symbol.

Compiler is error tolerant.  And that often means supporting broken code.  By creating a symbol here, later passes don't have to do things like ask "does this variable declarator have a symbol?" instead, they can just assume it does.  So this behavior is by design on the compiler side i believe.

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs (hunk 2 above)
Comment (genlu) [reply src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:135]: `CanBeReferencedByName` is actually true for the empty symbol. @333fred is this expected?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs (hunk 1 above)
Comment (genlu) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:10746]: Adding a space would change the context for completion, in which case no completion will trigger. Basically, what the current test does is to list all items at same context as `int` (which is treated as a local declaration and a local symbol with no name and of type int is generated, i.e. the one we need to filter out)

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs (hunk 2 above)
Comment (333fred) [reply src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:135]: Are these local symbols? If so, that is current expected behavior, but there's certainly an argument to be made that a local symbol with no name cannot be referenced by name :). The current behavior dates back to before roslyn went open source, so changing it is a risky proposition, but  I could see that as a valid change.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:135]: Everything else checks: 

```
                return SyntaxFacts.IsValidIdentifier(this.Name) &&
                    !SyntaxFacts.ContainsDroppedIdentifierCharacters(this.Name);
```

This really seems like the case for `locals/labels/aliases/rangevariablse`

Comment (genlu) [reply src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:135]: Yes, this is a local. Not sure what the bar is for back-compat, but I can file a bug to track this

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:135]: IMO, this should be a totally fine change, and def within the bar we have for changing behavior of existing apis.

Comment (genlu) [reply src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs:135]: Filed #47690

**Review by allisonchou — APPROVED** [review]
lgtm! 💯

**Pull request merged**